File: old/Confirmation.java
package net.volcano.jdacommands.model.menu;

import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.events.message.react.GenericMessageReactionEvent;
import net.volcano.jdautilities.constants.Colors;
import net.volcano.jdautilities.constants.Reactions;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

public class Confirmation extends Menu {
	
	private final String userId;
	private final CompletableFuture<Boolean> future;
	
	public Confirmation(String userId, String messageId) {
		this.userId = userId;
		future = new CompletableFuture<>();
		future.completeOnTimeout(false, 5, TimeUnit.MINUTES);
		future.whenComplete((v, t) -> {
			remove();
		});
		this.messageId = messageId;
		expiration = 1000 * 60 * 5;
	}
	
	@Override
	public void onReactionEvent(GenericMessageReactionEvent event) {
		
		if (event.getUserId().equals(userId)) {
			if (event.getReactionEmote().isEmoji()) {
				switch (event.getReactionEmote().getEmoji()) {
					case Reactions.YES -> {
						complete(true, event);
					}
					case Reactions.NO -> {
						complete(false, event);
					}
				}
			}
		}
		
	}
	
	private void complete(boolean bool, GenericMessageReactionEvent event) {
		future.complete(bool);
		event.getChannel()
				.retrieveMessageById(event.getMessageId())
				.flatMap(message -> {
					EmbedBuilder embedBuilder = new EmbedBuilder();
					embedBuilder.setColor(bool ? Colors.YES : Colors.NO);
					embedBuilder.setTitle("Action " + (bool ? "confirmed" : "cancelled"));
					embedBuilder.setDescription("No further action required.");
					return message.editMessageEmbeds(embedBuilder.build());
				}).queue();
	}
	
	public CompletableFuture<Boolean> getFuture() {
		return future;
	}
	
	public static EmbedBuilder getEmbed(String content) {
		EmbedBuilder embedBuilder = new EmbedBuilder();
		embedBuilder.setTitle("Confirm");
		embedBuilder.setDescription(content);
		embedBuilder.setColor(Colors.NOTES);
		embedBuilder.setFooter("You have 5 minutes to respond, after which this confirmation will automatically be denied.");
		return embedBuilder;
	}
	
}

File: src/main/java/net/volcano/jdacommands/model/interaction/Confirmation.kt
package net.volcano.jdacommands.model.interaction

import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.MessageBuilder
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.events.interaction.ButtonClickEvent
import net.dv8tion.jda.api.interactions.components.ActionRow
import net.dv8tion.jda.api.interactions.components.Button
import net.volcano.jdautils.constants.Colors
import java.util.concurrent.CompletableFuture

class Confirmation(
	val userId: String,
	val generator: (Boolean) -> Any?
) : InteractionListener(60L * 5L) {

	val future: CompletableFuture<Boolean> = CompletableFuture()

	private var expired = true

	override fun onInteraction(event: ButtonClickEvent) {

		if (event.user.id != userId) {
			val embedBuilder = EmbedBuilder()
			embedBuilder.setColor(Colors.ERROR)
			embedBuilder.setDescription("```diff\n- You cannot confirm or deny this action! -```")
			event.replyEmbeds(embedBuilder.build())
				.setEphemeral(true)
				.queue()
			return
		}

		if (event.interaction.componentId == "confirm") {
			confirm(true, event)
		} else if (event.interaction.componentId == "deny") {
			confirm(false, event)
		}
	}

	fun confirm(answer: Boolean, event: ButtonClickEvent) {
		expired = false
		event.message?.let { removeSelf(it) } ?: removeSelf()
		event.deferReply(true)
		future.complete(answer)

		val message = generator.invoke(answer).let {
			when (it) {
				is Message -> it
				is MessageBuilder -> it.build()
				is EmbedBuilder -> MessageBuilder().setEmbeds(it.build()).build()
				is MessageEmbed -> MessageBuilder().setEmbeds(it).build()
				else -> {
					val embedBuilder = EmbedBuilder()
					if (answer) {
						embedBuilder.setColor(Colors.SUCCESS)
						embedBuilder.setDescription("```diff\n+ Action confirmed. +```")
					} else {
						embedBuilder.setColor(Colors.ERROR)
						embedBuilder.setDescription("```diff\n- Action denied. -```")
					}
					MessageBuilder().setEmbeds(embedBuilder.build()).build()
				}
			}
		}

		event.reply(message)
			.setEphemeral(true)
			.queue()

		event.message?.let { removeSelf(it) } ?: removeSelf()
	}

	override fun destruct(message: Message) {
		message.delete().queue()
	}

	override fun onRemove() {
		if (expired && !future.isDone)
			future.complete(false)
	}

	companion object {

		fun message(content: String): Message {
			val embedBuilder = EmbedBuilder()
			embedBuilder.setTitle("Confirmation")
			embedBuilder.setDescription(content)
			embedBuilder.setColor(Colors.INFO)
			embedBuilder.setFooter("You have 5 minutes to respond, after which this confirmation will automatically be denied.")
			val messageBuilder = MessageBuilder()
			messageBuilder.setEmbeds(embedBuilder.build())
			messageBuilder.setActionRows(
				ActionRow.of(
					Button.success("confirm", "Confirm"),
					Button.danger("deny", "Deny")
				)
			)
			return messageBuilder.build()
		}

	}

}